File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example1/BookIterator.cs
namespace IteratorDesignPattern.Example1;

public class BookIterator : IIterator<Book>
{
    private readonly BookCollection _collection;
    private int _currentIndex = 0;

    public BookIterator(BookCollection collection)
    {
        _collection = collection;
    }

    public bool HasNext()
    {
        return _currentIndex < _collection.Count;
    }

    public Book Next()
    {
        if (!HasNext())
        {
            throw new InvalidOperationException("No more books in collection.");
        }

        return _collection[_currentIndex++];
    }

    public Book Current => _collection[_currentIndex];

    public void Reset()
    {
        _currentIndex = 0;
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example1/src/BookIterator.ts
import { Book } from "./Book";
import { BookCollection } from "./BookCollection";
import { IIterator } from "./Interfaces";

export class BookIterator implements IIterator<Book> {
  private _currentIndex = 0;

  constructor(private readonly _collection: BookCollection) {}

  hasNext(): boolean {
    return this._currentIndex < this._collection.count;
  }

  next(): Book {
    if (!this.hasNext()) {
      throw new Error("No more books in collection.");
    }
    return this._collection.getAt(this._currentIndex++);
  }

  current(): Book {
    return this._collection.getAt(this._currentIndex);
  }

  reset(): void {
    this._currentIndex = 0;
  }
}
